src/meta/src/barrier/mod.rs

```diff
@@ -815,61 +831,83 @@ where
-
-                // If no checkpoint, we can't notify collection completion
-                if *checkpoint {
-                    let mut uncommitted_states = checkpoint_control.get_uncommitted_states();
-                    if prev_epoch != INVALID_EPOCH {
-                        self.hummock_manager
-                            .commit_epoch(
-                                prev_epoch,
-                                uncommitted_states
-                                    .uncommitted_ssts
-                                    .into_iter()
-                                    .flatten()
-                                    .collect_vec(),
-                                uncommitted_states.uncommitted_work_ids,
-                            )
-                            .await?;
+                // See https://github.com/singularity-data/risingwave/issues/1251
+                let checkpoint = node.command_ctx.checkpoint;
+                let mut sst_to_worker: HashMap<HummockSstableId, WorkerId> = HashMap::new();
+                let mut synced_ssts: Vec<LocalSstableInfo> = vec![];
+                for resp in resps {
+                    let mut t: Vec<LocalSstableInfo> = resp
+                        .synced_sstables
+                        .iter()
+                        .cloned()
+                        .map(|grouped| {
+                            let sst = grouped.sst.expect("field not None");
+                            sst_to_worker.insert(sst.id, resp.worker_id);
+                            (grouped.compaction_group_id, sst)
+                        })
+                        .collect_vec();
+                    synced_ssts.append(&mut t);
+                }
+
+                // hummock_manager commit epoch.
+                if prev_epoch == INVALID_EPOCH {
+                    assert!(
+                        synced_ssts.is_empty(),
+                        "no sstables should be produced in the first epoch"
+                    );
+                } else if checkpoint {
+                    self.hummock_manager
+                        .commit_epoch(node.command_ctx.prev_epoch.0, synced_ssts, sst_to_worker)
+                        .await?;
+                } else {
+                    self.hummock_manager.update_current_epoch(prev_epoch)?;
+                    // if we collect a barrier(checkpoint = false),
+                    // we need to ensure that command is Plain and the notifier's checkpoint is
+                    // false
+                    assert_matches!(node.command_ctx.command, Command::Plain(_));
+                }
+
+                node.command_ctx.post_collect().await?;
+
+                // Notify about collected.
+                let mut notifiers = take(&mut node.notifiers);
+                notifiers.iter_mut().for_each(|notifier| {
+                    if !checkpoint {
+                        assert!(!notifier.bound_to_checkpoint_barrier())
                     }
```

Comment:
```suggestion                    
                    assert_eq!(checkpoint, notifier.bound_to_checkpoint_barrier())                   
```

Reply:
If checkpoint is true, `bound_to_checkpoint_barrier()` may be false.